src/liballoc/collections/btree/navigate.rs

```diff
@@ -120,56 +120,61 @@ impl<'a, K, V> Handle<NodeRef<marker::Immut<'a>, K, V, marker::Leaf>, marker::Ed
     }
 }
 
-impl<'a, K, V> Handle<NodeRef<marker::Mut<'a>, K, V, marker::Leaf>, marker::Edge> {
+impl<'a, K, V> Handle<NodeRef<marker::ValMut<'a>, K, V, marker::Leaf>, marker::Edge> {
     /// Moves the leaf edge handle to the next leaf edge and returns references to the
     /// key and value in between.
-    /// Unsafe for two reasons:
-    /// - The caller must ensure that the leaf edge is not the last one in the tree.
-    /// - Using the updated handle may well invalidate the returned references.
-    pub unsafe fn next_unchecked(&mut self) -> (&'a mut K, &'a mut V) {
+    /// Unsafe because the caller must ensure that the leaf edge is not the last one in the tree.
+    pub unsafe fn next_unchecked(&mut self) -> (&'a K, &'a mut V) {
         unsafe {
             let kv = replace(self, |leaf_edge| {
                 let kv = leaf_edge.next_kv();
                 let kv = unwrap_unchecked(kv.ok());
                 (ptr::read(&kv).next_leaf_edge(), kv)
             });
-            // Doing the descend (and perhaps another move) invalidates the references
-            // returned by `into_kv_mut`, so we have to do this last.
-            kv.into_kv_mut()
+            // Doing this last is faster, according to benchmarks.
+            kv.into_kv_valmut()
         }
     }
 
     /// Moves the leaf edge handle to the previous leaf and returns references to the
     /// key and value in between.
-    /// Unsafe for two reasons:
-    /// - The caller must ensure that the leaf edge is not the first one in the tree.
-    /// - Using the updated handle may well invalidate the returned references.
-    pub unsafe fn next_back_unchecked(&mut self) -> (&'a mut K, &'a mut V) {
+    /// Unsafe because the caller must ensure that the leaf edge is not the last one in the tree.
+    pub unsafe fn next_back_unchecked(&mut self) -> (&'a K, &'a mut V) {
         unsafe {
             let kv = replace(self, |leaf_edge| {
                 let kv = leaf_edge.next_back_kv();
                 let kv = unwrap_unchecked(kv.ok());
                 (ptr::read(&kv).next_back_leaf_edge(), kv)
             });
-            // Doing the descend (and perhaps another move) invalidates the references
-            // returned by `into_kv_mut`, so we have to do this last.
-            kv.into_kv_mut()
+            // Doing this last is faster, according to benchmarks.
+            kv.into_kv_valmut()
         }
     }
 }
 
+impl<'a, K, V> Handle<NodeRef<marker::Mut<'a>, K, V, marker::Leaf>, marker::Edge> {
+    /// Moves the leaf edge handle to the next leaf edge.
+    /// Unsafe because the caller must ensure that the leaf edge is not the last one in the tree.
+    pub unsafe fn move_next_unchecked(&mut self) {
+        let kv = unsafe { ptr::read(self).next_kv() };
+        let kv = unsafe { unwrap_unchecked(kv.ok()) };
+        *self = kv.next_leaf_edge();
```

Comment:
This looks like the `take_mut` pattern. That pattern is safe only if the code can never panic between the `ptr::read` and the `*self`. Could you ad a comment saying that, and also make sure that indeed `next_kv`, `unwrap_unchecked` and `next_leaf_edge` never panic?

Also maybe use `ptr::write` instead of `*self` -- it's more symmetric and means I don't have to check that this does not have a destructor.

Reply:
The `replace` function higher up, that a birdy whispered into my ear (I think it was Mark), seems to be an extended `take_mut` pattern. Originally in this `move_next_unchecked`, I reused `replace` with a void result, but that looked more complicated to me than writing it out with ptr::read/write. Now I put  in a `take` function very much like `take_mut::take`. But it has one disadvantage: the closure passed in doesn't and can't distinguish between safe code (like the `next_leaf_edge` above) and unsafe code (like `unwrap_unchecked`), because everything is already in an unsafe block.

But unlike `replace`, [`take_mut::take`](https://docs.rs/take_mut/0.2.2/take_mut/fn.take.html) is not marked unsafe. Some code shared in [RFC 1736](https://github.com/rust-lang/rfcs/pull/1736) is not marked unsafe either, and I seem to pick up that aborting the program because a closure panics is not undefined behaviour. I'm not even sure Mark or I slapped it on, so I'll be removing the unsafe unless you guys object.

Reply:
The old code needed `unsafe` because it did *not* abort.

Adding a safe `take` has been requested a few times I think, but there was never enough of a push for it. Actually having a user in libstd might change the game. ;)

Reply:
Relevant links: https://github.com/rust-lang/rfcs/issues/1508, https://github.com/rust-lang/rfcs/pull/1736.

Reply:
> The old code needed `unsafe` because it did _not_ abort.

Now I understand with "the old code" you mean the code in the current code, that is old. That is not currently the current code, because I removed the unsafe, but I'll put it back.

Regardless of whether it's unsafe or aborting, it is weird that Rust does not have something resembling `noexcept`. Unless this `take_mut` pattern is actually a bad idea. But I'm not going to try to touch it in btree right now.

Reply:
Ah yes I meant master, i.e., before your PR.

> Regardless of whether it's unsafe or aborting, it is weird that Rust does not have something resembling noexcept. Unless this take_mut pattern is actually a bad idea. But I'm not going to try to touch it in btree right now.

`noexcept` means we'd need an effect system -- because Rust is higher-order, adding a simple attribute is not enough, you have to be able to mark `fn` ptrs and methods in `trait` implementations as `noexcept`. It's a huge amount of compelxity for a niche usecase such as `take_mut`.

Reply:
Though I did manage to put detailed unsafe blocks in the closures passed to `replace`, by simply moving them out as a local variable.

Reply:
> Ah yes I meant master, i.e., before your PR.

Well, yes, but I didn't change that code, I just added and inlined one call of `replace`. So now I interpret your "The old code needed unsafe because it did not abort." as "The old code, that's still in there, needed and needs unsafe because it did not and does not abort".

Meanwhile I did "change that code" by removing the unsafe, but right now that's reverted.

Reply:
> also make sure that indeed [...], `unwrap_unchecked` and [...] never panic?

So I added an `unwrap_unchecked_unpanicky` that never does. The old `unwrap_unchecked` seems meant not to prevent panic but to optimize it away in a release build. One would indeed think the code to panic wastes resources, but I've always wondered how many pennies we save this way. Time to measure!

```
benchcmp old plain_unwrap_1 --threshold 5
 name                                           old ns/iter  plain_unwrap_1 ns/iter  diff ns/iter   diff %  speedup
 btree::map::insert_rand_10_000                 33           35                                 2    6.06%   x 0.94
 btree::map::iteration_1000                     4,027        1,133                         -2,894  -71.86%   x 3.55
 btree::map::iteration_100000                   498,075      294,756                     -203,319  -40.82%   x 1.69
 btree::map::iteration_20                       80           21                               -59  -73.75%   x 3.81
 btree::map::iteration_mut_1000                 3,898        1,019                         -2,879  -73.86%   x 3.83
 btree::map::iteration_mut_100000               476,445      275,236                     -201,209  -42.23%   x 1.73
 btree::map::iteration_mut_20                   78           21                               -57  -73.08%   x 3.71
 btree::set::difference_random_100_vs_100       745          686                              -59   -7.92%   x 1.09
 btree::set::difference_random_10k_vs_10k       186,910      163,967                      -22,943  -12.27%   x 1.14
 btree::set::difference_staggered_100_vs_100    765          693                              -72   -9.41%   x 1.10
 btree::set::difference_staggered_10k_vs_10k    73,799       64,744                        -9,055  -12.27%   x 1.14
 btree::set::intersection_random_100_vs_100     643          347                             -296  -46.03%   x 1.85
 btree::set::intersection_random_10k_vs_10k     163,212      120,081                      -43,131  -26.43%   x 1.36
 btree::set::intersection_staggered_100_vs_100  624          373                             -251  -40.22%   x 1.67
 btree::set::intersection_staggered_10k_vs_10k  60,801       34,355                       -26,446  -43.50%   x 1.77
 btree::set::is_subset_100_vs_100               589          302                             -287  -48.73%   x 1.95
 btree::set::is_subset_10k_vs_10k               58,620       29,608                       -29,012  -49.49%   x 1.98
```
I'm not kidding, I cleaned config.toml and rebuilt with the same result. Replacing the body of `unwrap_unchecked` with `val.unwrap()` boosts iteration performance considerably. This is not related to the panic-safe changes for this PR in navigate.rs, but the iterator code higher up in map.rs also does a lot of `unwrap_unchecked`.

Also replacing the body of `unwrap_unchecked_unpanicky`, thus violating safety preconditions, does not help performance.

But wait... am I just wrongly assuming that in the release builds benchmarked, the `if cfg!(debug_assertions)` code gets compiled away? No, making `unwrap_unchecked` identical to the new `unwrap_unchecked_unpanicky` does not change performance.

I conclude that this topic is unrelated to this PR. This PR (or actually, the PR where we introduced `replace`) needs an unpanicky `unwrap_unchecked` but not all `unwrap_unchecked` should be like that.

Reply:
Actually, I changed my mind a little. It does mean that I shouldn't sing the praise of `unwrap_unchecked` by listing a 2nd version in a central location, when it's only used in one file anyway. So moving it.

Reply:
Wait, the `unwrap` version is *faster* than the unsafe one that UBs on `None`? Wow, how on Earth...

Reply:
Well probably someone made sure `unwrap` generates excellent code, and this attempt at micro-optimization no longer pays off (if it ever did).

Reply:
It looks more reasonable 'for a [standalone microbenchmark](https://github.com/ssomers/rust_bench_quickies): 
```
test option_unwrap::unwrap                               ... bench:         104 ns/iter (+/- 3)
test option_unwrap::unwrap_or_else_unreachable           ... bench:          70 ns/iter (+/- 1)
test option_unwrap::unwrap_or_else_unreachable_unchecked ... bench:          55 ns/iter (+/- 0)
```
Here it's only weird that the middle one (`unwrap_or_else(|| unreachable!()))`) appears faster than plain `unwrap`. That difference disappears if we make the option contents big (the option contents are a simple reference in the numbers above), while the `unreachable_unchecked` code stays twice as fast.

To be continued in another PR.